assume/units/building.py

```diff
@@ -0,0 +1,401 @@
+# SPDX-FileCopyrightText: ASSUME Developers
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+import ast
+import logging
+
+import pandas as pd
+import pyomo.environ as pyo
+from pyomo.opt import (
+    SolverFactory,
+    SolverStatus,
+    TerminationCondition,
+    check_available_solvers,
+)
+
+from assume.common.base import SupportsMinMax
+from assume.common.market_objects import MarketConfig, Orderbook
+from assume.common.utils import get_products_index
+from assume.units.dsm_load_shift import DSMFlex
+from assume.units.dst_components import (
+    create_boiler,
+    create_ev,
+    create_heatpump,
+    create_thermal_storage,
+)
+
+SOLVERS = ["gurobi", "glpk", "cbc", "cplex"]
+
+logger = logging.getLogger(__name__)
+
+# Mapping of component type identifiers to their respective classes
+building_components = {
+    "heatpump": create_heatpump,
+    "boiler": create_boiler,
+    "thermal_storage": create_thermal_storage,
+    "ev": create_ev,
+}
+
+
+class Building(SupportsMinMax, DSMFlex):
+    """
+    The Building class represents a building unit in the energy system.
+
+    Args:
+        id (str): The unique identifier of the unit.
+        unit_operator (str): The operator of the unit.
+        bidding_strategies (dict): The bidding strategies of the unit.
+        technology (str): The technology of the unit.
+        node (str): The node of the unit.
+        index (pd.DatetimeIndex): The index for the data of the unit.
+        location (tuple[float, float]): The location of the unit.
+        components (dict[str, dict]): The components of the unit such as heat pump, electric boiler, and thermal storage.
+        objective (str): The objective of the unit, e.g. minimize expenses ("minimize_expenses").
+    """
+
+    def __init__(
+        self,
+        id: str,
+        unit_operator: str,
+        bidding_strategies: dict,
+        technology: str = "building",
+        node: str = "node0",
+        index: pd.DatetimeIndex = None,
+        location: tuple[float, float] = (0.0, 0.0),
+        components: dict[str, dict] = None,
+        flexibility_measure: str = "max_load_shift",
+        demand: float = 0,
+        objective: str = None,
+        **kwargs,
+    ):
+        super().__init__(
+            id=id,
+            unit_operator=unit_operator,
+            technology=technology,
+            bidding_strategies=bidding_strategies,
+            index=index,
+            node=node,
+            location=location,
+            **kwargs,
+        )
+
+        self.electricity_price = self.forecaster["price_EOM"]
+        self.natural_gas_price = self.forecaster["fuel_price_natural_gas"]
```

Comment:
`fuel_price_natural_gas` needs to be called `fuel_price_natural gas` since in the input file it is written `natural gas` too. Or the csv_loader gets adapted such that it replaces whitespaces with "_".